### PR TITLE
ENH Adds _safe_softmax_last_dim for multiheadattention 

### LINKED ISSUE #41508: nn.MultiheadAttention causes gradients to become NaN under some use cases
## 🐛 Bug

Using key_padding_mask and attn_mask with nn.MultiheadAttention causes gradients to become NaN under some use cases.

## To Reproduce

Backwards pass through nn.MultiheadAttention layer where the forward pass used:
1. attn_mask limiting context in both directions (e.g. bucketed attention)
2. key_padding_mask where there is padding for at least one sequence (and there is also at least one valid entry for every sequence, as expected)
3. The timesteps that were masked are not used to calculate the loss

import torch
torch.manual_seed(0)
'''Create attention layer'''
attn = torch.nn.MultiheadAttention(embed_dim=1, num_heads=1)
'''Create dummy input'''
x = torch.rand(3, 2, 1)
'''Padding mask, second sequence can only see first embedding'''
key_padding_mask = torch.as_tensor([[False, False, False], [False, True, True]], dtype=torch.bool)
'''Attention mask, bucketing attention to current and previous time steps'''
attn_mask = torch.as_tensor([[0., float('-inf'), float('-inf')], [0., 0., float('-inf')], [float('-inf'), 0., 0.]])
'''Generate attention embedding'''
output, scores = attn(x, x, x, key_padding_mask=key_padding_mask, attn_mask=attn_mask)
print("scores")
print(scores)
'''Create a dummy loss, only use the first embedding which is defined for all sequences'''
loss = output[0, :].sum()
print("loss")
print(loss)
'''Backwards pass and gradients'''
loss.backward()
print("grads")
for n, p in attn.named_parameters():
    print(n, p.grad)

> scores
> tensor([[[1.0000, 0.0000, 0.0000],
>          [0.4468, 0.5532, 0.0000],
>          [0.0000, 0.5379, 0.4621]],
>         [[1.0000, 0.0000, 0.0000],
>         [1.0000, 0.0000, 0.0000],
>          [   nan,    nan,    nan]]], grad_fn=<DivBackward0>)
> loss
> tensor(0.0040, grad_fn=<SumBackward0>)
> grads
> in_proj_weight tensor([[nan],
>         [nan],
>         [nan]])
> in_proj_bias tensor([nan, nan, nan])
> out_proj.weight tensor([[nan]])
> out_proj.bias tensor([2.])

## Expected behavior

Gradients should not be NaN

## Environment

PyTorch version: 1.5.1
Is debug build: No
CUDA used to build PyTorch: None

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.5.1
[conda] blas                      1.0                         mkl
[conda] cpuonly                   1.0                           0    pytorch
[conda] mkl                       2020.1                      217
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.1.0            py37h23d657b_0
[conda] mkl_random                1.1.1            py37h0573a6f_0
[conda] numpy                     1.18.5           py37ha1c710e_0
[conda] numpy-base                1.18.5           py37hde5b4d6_0
[conda] pytorch                   1.5.1               py3.7_cpu_0  [cpuonly]  pytorch

Also fails when using GPU.

### LINKED ISSUE #55056: [FR] Safe softmax
## 🚀 Feature
Expand softmax to support "safe softmax" behavior (i.e. output 0 instead of NaN when called with no inputs).

## Motivation
When softmax is called with no inputs, the output is NaN due to computation of 0/0. In certain cases, an output of 0 may be preferred instead.

For example, due to masking in MHA, a batch may include a sentence that is "masked out" / consists entirely of padding. While this case will currently result in NaNs, leading to training divergence, it is often preferred to effectively ignore these cases so training can continue. See https://github.com/pytorch/pytorch/issues/41508 and https://github.com/pytorch/pytorch/issues/40932 for more info.

## Pitch
Add a new, optional `eps` parameter with default `None` to the various softmax forms:

```python
from torch import nn
from torch.nn import functional as F

m = nn.Softmax(dim=-1, eps=1e-5)
m2 = nn.Softmax2d(eps=1e-5)
output = F.softmax(..., eps=1e-5)
```

If specified, `eps` will be included in the denominator of the softmax computation, resulting in `0/(0+eps) = 0` instead of `0/0 = NaN` for the no input case. See https://github.com/pytorch/pytorch/issues/41508#issuecomment-775030992.

## Alternatives
Leave as-is; calling softmax with no inputs will unconditionally result in NaNs. It is left to users to ensure this doesn't happen if it's undesired, and failure to handle this correctly will still be considered "user error".

### PR DESCRIPTION
**Pull request by thomasjpfan**
Fixes https://github.com/pytorch/pytorch/issues/41508
Fixes https://github.com/pytorch/pytorch/issues/55056

1. This PR adds a private `_zero_if_all_neg_inf` kwarg to `softmax` that sets the softmax to 0 when `-inf`. Another option would be to add a new `_safe_softmax` operator. 

2. For CPU kernels, the `_zero_if_all_neg_inf` is passed to the kernel as a boolean option. For CUDA, `_zero_if_all_neg_inf` is passed down and the dispatched to different templates.

3. I ran the following CPU/GPU benchmark script and this PR does not add any additional overhead to softmax:

<details>
 <summary>Benchmark script</summary>

```python
from itertools import product

import torch
import torch.nn as nn
import torch.nn.functional as F
import time
import statistics

torch.manual_seed(0)
MS_PER_SECOND = 1000


def _time(device):
    if device == "cuda":
        torch.cuda.synchronize()
    return time.perf_counter() * MS_PER_SECOND


devices = ["cuda", "cpu"]
n_runs = 30
device_Ns = [
    ("cuda", 64),
    ("cuda", 128),
    ("cuda", 256),
    ("cuda", 512),
    ("cuda", 1024),
    ("cpu", 64),
    ("cpu", 128),
    ("cpu", 256),
]
dims = [0, 1, 2]

for (device, N), dim in product(device_Ns, dims):
    total_fwd_times = []

    for _ in range(n_runs):
        input = torch.randn(N, 1000, 1000, device=device)

        # forward
        start = _time(device)
        F.softmax(input, dim=dim)
        total_fwd_times.append(_time(device) - start)

    fwd_avg = statistics.mean(total_fwd_times)
    fwd_std = statistics.stdev(total_fwd_times)
    print(
        f"input size({N}, 1000, 1000), dim={dim}, device={device}, "
        f"fwd: {fwd_avg:.2f} +/- {fwd_std:.2f}(ms)"
    )

```

</details>

### CODE REVIEW

**Comment by thomasjpfan** [comment]
I updated this PR by adding a new private operator: `_safe_softmax_last_dim`. Since MHA only uses the last dimension for softmax, I restricted this PR to only work with the last dimension. This reduces the number of changes in the CPU kernels.

For the CUDA kernels, this PR updates `SoftMaxForwardEpilogue` to template out `_zero_if_all_neg_inf`, which is used to compute the softmax.

**Comment by thomasjpfan** [comment]
Comparing `_safe_softmax_last_dim` to `softmax(dim=-1)` with a tensor with `-inf`.

<details>
 <summary>Benchmark script</summary>

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
import time
import statistics

torch.manual_seed(0)
MS_PER_SECOND = 1000


def _time(device):
    if device == "cuda":
        torch.cuda.synchronize()
    return time.perf_counter() * MS_PER_SECOND


devices = ["cuda", "cpu"]
n_runs = 30
device_Ns = [
    ("cuda", 64),
    ("cuda", 128),
    ("cuda", 256),
    ("cuda", 512),
    ("cuda", 1024),
    ("cpu", 64),
    ("cpu", 128),
]

for device, N in device_Ns:
    total_softmax_times = []
    total_safe_softmax_times = []

    for _ in range(n_runs):
        input = torch.randn(N, 1000, 1000, device=device)
        input[N//2:] = float("-inf")

        # forward
        start = _time(device)
        input.softmax(dim=-1)
        total_softmax_times.append(_time(device) - start)

        # forward
        start = _time(device)
        input._safe_softmax_last_dim()
        total_safe_softmax_times.append(_time(device) - start)

    soft_fwd_avg = statistics.mean(total_softmax_times)
    soft_fwd_std = statistics.stdev(total_softmax_times)

    safe_fwd_avg = statistics.mean(total_safe_softmax_times)
    safe_fwd_std = statistics.stdev(total_safe_softmax_times)
    print(
        f"input size({N}, 1000, 1000), device={device}, "
        f"soft_fwd: {soft_fwd_avg:.2f} +/- {soft_fwd_std:.2f}(ms) "
        f"safe_fwd: {safe_fwd_avg:.2f} +/- {safe_fwd_std:.2f}(ms) "
    )


```

</details>

I get:

```
input size(64, 1000, 1000), device=cuda, soft_fwd: 0.63 +/- 0.05(ms) safe_fwd: 0.62 +/- 0.00(ms) 
input size(128, 1000, 1000), device=cuda, soft_fwd: 1.23 +/- 0.05(ms) safe_fwd: 1.22 +/- 0.00(ms) 
input size(256, 1000, 1000), device=cuda, soft_fwd: 2.45 +/- 0.10(ms) safe_fwd: 2.43 +/- 0.00(ms) 
input size(512, 1000, 1000), device=cuda, soft_fwd: 4.88 +/- 0.18(ms) safe_fwd: 4.86 +/- 0.01(ms) 
input size(1024, 1000, 1000), device=cuda, soft_fwd: 9.75 +/- 0.31(ms) safe_fwd: 9.71 +/- 0.01(ms) 
input size(64, 1000, 1000), device=cpu, soft_fwd: 29.37 +/- 1.28(ms) safe_fwd: 14.65 +/- 0.71(ms) 
input size(128, 1000, 1000), device=cpu, soft_fwd: 60.92 +/- 1.57(ms) safe_fwd: 29.77 +/- 1.03(ms)
```

The safe softmax does not add any time to the `cuda` implementation and reduces the runtime for `cpu`.

**Review by jbschlosser — CHANGES_REQUESTED** [review]
As discussed offline, we have decided not to move forward with this approach, but rather to wait for a better fix involving correct masking semantics.

File: torch/nn/functional.py [hunk 1]
```diff
@@ -4840,7 +4840,7 @@ def _scaled_dot_product_attention(
     attn = torch.bmm(q, k.transpose(-2, -1))
     if attn_mask is not None:
         attn += attn_mask
-    attn = softmax(attn, dim=-1)
+    attn = attn._safe_softmax_last_dim()
```
[the remark this answers is not in the document]
Comment (rodosingh) [reply torch/nn/functional.py:4843]: @jbschlosser and @ngimel I found out that using `nan_to_num(0.0)` doesn't solve the issue as after the **first forward pass**, `Q`, `K`, and `V` (of course) becomes `nan` (tried and tested).

So without even going for replacing `1`s with `-inf` as done in `.masked_fill(tensor, float("-inf"))`, we can just multiply `1-attn_mask` (as `PyTorch` convention of masking is different from that of `Hugging-Face`) with `softmax(Q*K^T)`, and that solves the problem. Again we just need to do `.masked_fill(tensor, 1.0)`.

Here is the snippet of code that describes what I explained above:
```python3
def _scaled_dot_product_attention(
    q: Tensor,
    k: Tensor,
    v: Tensor,
    attn_mask: Optional[Tensor] = None,
    dropout_p: float = 0.0,
) -> Tuple[Tensor, Tensor]:
    r"""
    Computes scaled dot product attention on query, key and value tensors, using
    an optional attention mask if passed, and applying dropout if a probability
    greater than 0.0 is specified.
    Returns a tensor pair containing attended values and attention weights.

    Args:
        q, k, v: query, key and value tensors. See Shape section for shape details.
        attn_mask: optional tensor containing mask values to be added to calculated
            attention. May be 2D or 3D; see Shape section for details.
        dropout_p: dropout probability. If greater than 0.0, dropout is applied.

    Shape:
        - q: :math:`(B, Nt, E)` where B is batch size, Nt is the target sequence length,
            and E is embedding dimension.
        - key: :math:`(B, Ns, E)` where B is batch size, Ns is the source sequence length,
            and E is embedding dimension.
        - value: :math:`(B, Ns, E)` where B is batch size, Ns is the source sequence length,
            and E is embedding dimension.
        - attn_mask: either a 3D tensor of shape :math:`(B, Nt, Ns)` or a 2D tensor of
            shape :math:`(Nt, Ns)`.

        - Output: attention values have shape :math:`(B, Nt, E)`; attention weights
            have shape :math:`(B, Nt, Ns)`
    """
    B, Nt, E = q.shape
    q = q / math.sqrt(E)
    # (B, Nt, E) x (B, E, Ns) -> (B, Nt, Ns)
    # if attn_mask is not None:
    #     attn = torch.baddbmm(attn_mask, q, k.transpose(-2, -1))
    # else:
    attn = torch.bmm(q, k.transpose(-2, -1))

    attn = F.softmax(attn, dim=-1)
    if attn_mask is not None:
        attn = attn * (1-attn_mask)
    if dropout_p > 0.0:
        attn = F.dropout(attn, p=dropout_p)
    # (B, Nt, Ns) x (B, Ns, E) -> (B, Nt, E)
    output = torch.bmm(attn, v)
    return output, attn
```
#### Detailed explanation and cause of issue is explained [here](https://github.com/pytorch/pytorch/issues/41508#issuecomment-1723119580) at #41508.
